File: infrastructures/rest-ddd/python/src/infrastructure/persistence/write/base_write_entity.py
"""
Base Write Entity - For Commands (CREATE, UPDATE, DELETE)
Used when business logic executes against data.
Contains all fields needed for business operations.
"""
from datetime import datetime
from typing import Optional
from abc import ABC


class BaseWriteEntity(ABC):
    """
    Abstract base class for all Write Entities (Command side).
    
    Rules:
    - MUST be in /infrastructure/persistence/write/
    - MUST suffix with 'WriteEntity'
    - Contains ORM/database attributes
    - Used for commands that modify data
    - Contains version for optimistic locking
    """
    
    def __init__(
        self,
        p_id: Optional[str] = None,
        p_version: Optional[int] = None,
        p_created_at: Optional[datetime] = None,
        p_updated_at: Optional[datetime] = None
    ):
        self._id = p_id
        self._version = p_version or 1
        self._created_at = p_created_at or datetime.utcnow()
        self._updated_at = p_updated_at or datetime.utcnow()
    
    @property
    def id(self) -> Optional[str]:
        return self._id
    
    @id.setter
    def id(self, p_value: str) -> None:
        self._id = p_value
    
    @property
    def version(self) -> int:
        return self._version
    
    @version.setter
    def version(self, p_value: int) -> None:
        self._version = p_value
    
    @property
    def created_at(self) -> datetime:
        return self._created_at
    
    @property
    def updated_at(self) -> datetime:
        return self._updated_at
    
    @updated_at.setter
    def updated_at(self, p_value: datetime) -> None:
        self._updated_at = p_value
    
    def increment_version(self) -> None:
        """Increment version for optimistic locking."""
        self._version += 1
        self._updated_at = datetime.utcnow()

File: infrastructures/rest-ddd/csharp/src/Infrastructure/Persistence/Write/BaseWriteEntity.cs
/**
 * Base Write Entity - For Commands (CREATE, UPDATE, DELETE)
 * Used when business logic executes against data.
 * Contains all fields needed for business operations.
 */

using System;

namespace Infrastructure.Persistence.Write
{
    public abstract class BaseWriteEntity
    {
        /**
         * Abstract base class for all Write Entities (Command side).
         * 
         * Rules:
         * - MUST be in /Infrastructure/Persistence/Write/
         * - MUST suffix with 'WriteEntity'
         * - Contains ORM/database attributes
         * - Used for commands that modify data
         * - Contains version for optimistic locking
         */

        protected string? _id;
        protected int _version;
        protected DateTime _createdAt;
        protected DateTime _updatedAt;

        protected BaseWriteEntity()
        {
            _version = 1;
            _createdAt = DateTime.UtcNow;
            _updatedAt = DateTime.UtcNow;
        }

        public string? Id
        {
            get => _id;
            set => _id = value;
        }

        public int Version
        {
            get => _version;
            set => _version = value;
        }

        public DateTime CreatedAt => _createdAt;

        public DateTime UpdatedAt
        {
            get => _updatedAt;
            set => _updatedAt = value;
        }

        /**
         * Increment version for optimistic locking.
         */
        public void IncrementVersion()
        {
            _version++;
            _updatedAt = DateTime.UtcNow;
        }
    }
}
